BinnedCountPredictionModel.py

```diff
@@ -0,0 +1,468 @@
+            'model_state_dict': model.state_dict(),
+            'optimizer_state_dict': optimizer.state_dict(),
+            'loss': test_loss
+        }
+        torch.save(state_dict, "bestmore.pt")
+        
+    # Stop training if we have not improved after X epochs
+    best_epoch = [i for i,j in enumerate(epoch_test_losses) if j == min(epoch_test_losses)][-1]
+    offset = epoch - best_epoch
+    if offset >= patience:
+        break
+        
+    #-l gpu_type = gpu100, v100
+#Load best model
+checkpoint = torch.load(
+    "best.pt",
+    map_location=lambda storage, loc: storage
+)
+best_epoch = checkpoint["epoch"]
+#model = Net(filter_sizes, fcl_layers).to(device)
+model = ResNet(fcl_layers, dr = dropout, output_size = output_size).to(device)
+model.load_state_dict(checkpoint["model_state_dict"])
+#predict on test dataset
+topk = (1,2)
+
+
+model.eval()
+with torch.no_grad():
+    y_true = []
+    y_pred = []
+    batch_loss = []
+    accuracy = {k: [] for k in topk}
+    # split test data into batches
+
+    valid_batches_per_epoch = int(X_test.shape[0] / valid_batch_size) 
+    batch_group_generator = tqdm(
+        test_loader, 
+        total=valid_batches_per_epoch, 
+        leave=True
+    )
+    
+    for (x, y) in batch_group_generator:
+        # Converting to torch tensors and moving to GPU
+        inputs = x.float().to(device)
+        lightning_counts = y.long().to(device)
+        # get output from the model, given the inputs
+        pred_lightning_counts = model(inputs)
+        # get loss for the predicted output
+        loss = test_criterion(pred_lightning_counts, lightning_counts.squeeze(-1))
+        batch_loss.append(loss.item())
+        # compute the accuracy
+        acc = torch_accuracy(pred_lightning_counts, lightning_counts, topk = topk)
+        for i,k in enumerate(topk):
+            accuracy[k] += [acc[i]]
+        
+        y_true.append(lightning_counts.squeeze(-1))
+        # Taking the top-1 answer here, but here is where we could compute the average predicted rather than take top-1
+        y_pred.append(torch.argmax(pred_lightning_counts, 1))
+
+y_true = torch.cat(y_true, axis = 0)
+y_pred = torch.cat(y_pred, axis = 0)
+#save out to csv for analysis
+#save out batch loss and accuracy
+print("batch_loss",np.mean(batch_loss),"accuracy",np.mean(accuracy))
+y_true = y_true.cpu().numpy()
+y_pred = y_pred.cpu().numpy()
+np.savetxt("y_true.csv", y_true, delimiter=",")
+np.savetxt("y_pred.csv", y_pred, delimiter=",")
+print("y_true",y_true)
+print("y_pred",y_pred)
+print("val_loss", np.mean(batch_loss))
+for k in topk:
+    print(f"top-{k} {np.mean(accuracy[k])}")
+for label in list(set(y_true)):
+    c = (y_true == label)
+    print(label, (y_true[c] == y_pred[c]).mean())
+list(set(y_pred))
+import seaborn as sn
+import pandas as pd
+cm = metrics.confusion_matrix(y_true, y_pred, normalize = 'true')
+df_cm = pd.DataFrame(cm, index = sorted(list(set(y_true))), columns = sorted(list(set(y_true))))
```

Comment:
Save out your confusion matrix and predictions to csv to facilitate further analysis.